### ISSUE TITLE
TLS 1.3: SRV: PSK resumption is not selected when ephemeral mode is available

### ISSUE THREAD

**Issue by lpy4105** [Mbed-TLS/mbedtls#7069]
If the client would like a **psk resumption** and **ephemeral** mode is also available, the server, according to current key exchange mode preference, would select **ephemeral** mode instead of doing a **psk resumption**.  That is, if the client sends correct `pre_shared_key`, `pre_shared_key_exchange_modes`, and `key_share` extensions, the server will select the **ephemeral** mode even if the resumption psk identity in `pre_shared_key` ext is matched.

I think this should be fixed as the client send valid extensions for psk resumption, the server should do a psk resumption rather than ephemeral. At least, it causes problem in some test cases [here](https://github.com/Mbed-TLS/mbedtls/pull/6616#discussion_r1070857419).

There is already a [possible solution  in PR6945](https://github.com/Mbed-TLS/mbedtls/pull/6945/commits/dadeb20383956f6b8654fce1501ab2d572f09058), but it is still worth discussing if we need to change the priority of key exchange mode alternatively as mentioned in https://github.com/Mbed-TLS/mbedtls/pull/6945#discussion_r1100923100

**Comment by ronald-cron-arm** [comment]
I'd say that if the client offers pure psk and ephemeral as possible key exchange modes then quite naturally it would offer psk_ephemeral as well. As psk_ephemeral is the preferred key exchange mode our TLS 1.3 server would select that mode and actually use the psk..

**Comment by lpy4105** [comment]
> I'd say that if the client offers pure psk and ephemeral as possible key exchange modes then quite naturally it would offer psk_ephemeral as well.

Yes, but the client might offer arbitrary value in `psk_key_exchange_modes` extension for whatever reason. It's fine for the first flight, but for the second flight, the question would become to "Do we need to insist on `ephemeral` over `pure psk` even if there is a valid `ticket identity` in `pre_shared_key` extension for resumption?"

**Comment by yuhaoth** [comment]
> I'd say that if the client offers pure psk and ephemeral as possible key exchange modes then quite naturally it would offer psk_ephemeral as well.

I do not think so. In theory, psk and ephemeral mode can be configured, it is not forbidden by RFC.  

GNUTLS can be configured as psk and ephemeral with `NONE:+VERS-TLS1.3:+CIPHER-ALL:+MAC-ALL:+SIGN-ALL:+GROUP-ALL:+PSK`. I did not test the behaviors of resumption, I think we can try that.

**Comment by lpy4105** [comment]
> GNUTLS can be configured as psk and ephemeral with NONE:+VERS-TLS1.3:+CIPHER-ALL:+MAC-ALL:+SIGN-ALL:+GROUP-ALL:+PSK. I did not test the behaviors of resumption, I think we can try that.

I've tested that GNUTLS would do the resumption for a client that claims only support pure PSK and ephemeral. Please check the cases added in: https://github.com/lpy4105/mbedtls/commit/62b25b64b3d6cbab3c9294d0edea16246f360575
